ludwig/hyperopt/run.py

```diff
@@ -423,7 +429,80 @@ def log_warning_if_all_grid_type_parameters(hyperopt_parameter_config: Dict[str,
     )
 
 
-def update_hyperopt_params_with_defaults(hyperopt_params):
+def update_or_set_max_concurrent_trials(executor_config: dict, backend: Backend) -> None:
+    """Datasets read tasks request 0.5 CPUs and all transformation tasks request 1 CPU, so if there are no cores
+    available, trials won't be able to run.
+
+    Set max_concurrent_trials in the hyperopt executor to ensure CPU resources are available for Ray Dataset related
+    tasks.
+    """
+    cpu_resources_per_trial = executor_config.get(CPU_RESOURCES_PER_TRIAL, 1)
+    num_samples = executor_config.get(NUM_SAMPLES)
+    max_concurrent_trials = executor_config.get(MAX_CONCURRENT_TRIALS)
+
+    # Default to num_samples if max_concurrent_trials isn't set, otherwise use the value from the config
+    is_auto = max_concurrent_trials == AUTO
+    max_concurrent_trials = num_samples if max_concurrent_trials == AUTO else max_concurrent_trials
+
+    # Max_concurrent_trials set to None, no need to infer
+    if not max_concurrent_trials:
```

Comment:
We should also add a return in cases when `max_concurrent_trials` is set to an integer value in the executor config by the user.